pkg/3scale/amp/operator/backend_options_provider.go

```diff
@@ -0,0 +1,144 @@
+		apimanagerSpec: apimanagerSpec,
+		namespace:      namespace,
+		client:         client,
+		backendOptions: component.NewBackendOptions(),
+		secretSource:   NewSecretSource(client, namespace),
+	}
+}
+
+func (o *OperatorBackendOptionsProvider) GetBackendOptions() (*component.BackendOptions, error) {
+	o.backendOptions.AppLabel = *o.apimanagerSpec.AppLabel
+	o.backendOptions.TenantName = *o.apimanagerSpec.TenantName
+	o.backendOptions.WildcardDomain = o.apimanagerSpec.WildcardDomain
+
+	err := o.setSecretBasedOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	o.setResourceRequirementsOptions()
+	o.setReplicas()
+
+	err = o.backendOptions.Validate()
+	return o.backendOptions, err
+}
+
+func (o *OperatorBackendOptionsProvider) setSecretBasedOptions() error {
+	cases := []struct {
+		field       *string
+		secretName  string
+		secretField string
+		defValue    string
+	}{
+		{
+			&o.backendOptions.SystemBackendUsername,
+			component.BackendSecretInternalApiSecretName,
+			component.BackendSecretInternalApiUsernameFieldName,
+			component.DefaultSystemBackendUsername(),
+		},
+		{
+			&o.backendOptions.SystemBackendPassword,
+			component.BackendSecretInternalApiSecretName,
+			component.BackendSecretInternalApiPasswordFieldName,
+			component.DefaultSystemBackendPassword(),
+		},
+		{
+			&o.backendOptions.ServiceEndpoint,
+			component.BackendSecretBackendListenerSecretName,
+			component.BackendSecretBackendListenerServiceEndpointFieldName,
+			component.DefaultBackendServiceEndpoint(),
+		},
+		{
+			&o.backendOptions.RouteEndpoint,
+			component.BackendSecretBackendListenerSecretName,
+			component.BackendSecretBackendListenerRouteEndpointFieldName,
+			fmt.Sprintf("https://backend-%s.%s", *o.apimanagerSpec.TenantName, o.apimanagerSpec.WildcardDomain),
+		},
+		{
+			&o.backendOptions.StorageURL,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisStorageURLFieldName,
+			component.DefaultBackendRedisStorageURL(),
+		},
+		{
+			&o.backendOptions.QueuesURL,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisQueuesURLFieldName,
+			component.DefaultBackendRedisQueuesURL(),
+		},
+		{
+			&o.backendOptions.StorageSentinelHosts,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisStorageSentinelHostsFieldName,
+			"",
+		},
+		{
+			&o.backendOptions.StorageSentinelRole,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisStorageSentinelRoleFieldName,
+			"",
+		},
+		{
+			&o.backendOptions.QueuesSentinelHosts,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisQueuesSentinelHostsFieldName,
+			"",
+		},
+		{
+			&o.backendOptions.QueuesSentinelRole,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisQueuesSentinelRoleFieldName,
+			"",
+		},
+	}
+
+	for _, option := range cases {
```

Comment:
In setSecretBasedOptions how expected mandatory fields are going to be managed?
The code only seems to manage fields that have defaults.
I understand this is still pending implementation right and has not been done because it's not been the case here yet that we have an option in that situation.

Reply:
`SecretSource` struct has the following public methods available:

```
func (s *SecretSource) FieldValue(secretName, fieldName string, def string) (*string, error)
func (s *SecretSource) FieldValueFromRequiredSecret(secretName, fieldName string, def string) (*string, error)
func (s *SecretSource) RequiredFieldValue(secretName, fieldName string) (*string, error)
func (s *SecretSource) RequiredFieldValueFromRequiredSecret(secretName, fieldName string) (*string, error)
```

I think all cases are covered. For the backend options specific case, the secret existence is not mandatory and any field existence is mandatory. Thus `FieldValue` method is being used.

Reply:
:+1: . I thought you were trying to make the setSecretBasedOptions valid for all scenarios but for what I've understood for other comments that is specific code for the backend_options_provider and thje structure of the setSecretBasedOptions for other components might change a lot (using different structs, different methods of the SecretSource class, ...) :+1: